src/redux.js

```diff
@@ -898,3 +898,124 @@ export function getCurrentColumnIsSelectedLabel(state) {
 export function getCurrentColumnIsSelectedFeature(state) {
   return state.selectedFeatures.includes(state.currentColumn);
 }
+
+/* Returns an object with information for the CrossTab UI.
+ *
+ * Here is an example result:
+ *
+ *  {
+ *    results: [
+ *      {
+ *        featureValues: ["1", "1"],
+ *        labelCounts: { yes: 2, no: 1 },
+ *        labelPercents: { yes: 67, no: 33 }
+ *      },
+ *      {
+ *        featureValues: ["0", "0"],
+ *        labelCounts: { yes: 25, no: 42 },
+ *        labelPercents: { yes: 37, no: 63 }
+ *      },
+ *      {
+ *        featureValues: ["1", "0"],
+ *        labelCounts: { yes: 6, no: 5 },
+ *        labelPercents: { yes: 55, no: 45 }
+ *      },
+ *      {
+ *        featureValues: ["0", "1"],
+ *        labelCounts: { no: 2, yes: 2 },
+ *        labelPercents: { no: 50, yes: 50 }
+ *      }
+ *    ],
+ *    uniqueLabelValues: ["yes", "no"],
+ *    featureNames: ["caramel", "crispy"],
+ *    labelName: "delicious?"
+ *  }
+ *
+ */
+
+export function getCrossTabData(state) {
+  if (!state.labelColumn || state.selectedFeatures.length <= 0) {
+    return null;
+  }
+
+  var results = [];
+
+  // For each row of data, determine whether we have found a new or existing
+  // combination of feature values.  If new, then add a new entry to our results
+  // array.  Then record or increment the count for the corresponding label
+  // value.
+
+  for (let row of state.data) {
+    var featureValues = [];
+    for (let selectedFeature of state.selectedFeatures) {
+      featureValues.push(row[selectedFeature]);
+    }
+
+    var existingEntry = results.find(result => {
+      return areArraysEqual(result.featureValues, featureValues);
+    });
+
+    if (!existingEntry) {
+      existingEntry = {
+        featureValues,
+        labelCounts: { [row[state.labelColumn]]: 1 }
+      };
+      results.push(existingEntry);
+    } else {
+      if (!existingEntry.labelCounts[row[state.labelColumn]]) {
+        existingEntry.labelCounts[row[state.labelColumn]] = 1;
+      } else {
+        existingEntry.labelCounts[row[state.labelColumn]]++;
+      }
+    }
+  }
+
+  // Now that we have all the counts of label values, we can determine the
+  // corresponding percentage values.
+
+  for (let result of results) {
+    let totalCount = 0;
+    for (let labelCount of Object.values(result.labelCounts)) {
+      totalCount += labelCount;
+    }
+    result.labelPercents = {};
+    for (let key of Object.keys(result.labelCounts)) {
+      result.labelPercents[key] = Math.round(
+        (result.labelCounts[key] / totalCount) * 100
+      );
+    }
+  }
```

Comment:
Have you tried this with a big dataset? I have a general concern about performance in scenarios where we go through a whole column; not blocking for this, just curious.

Reply:
Good point.  I suspect we have a number of perf bottlenecks and we will need to understand how big the incoming datasets can get, but I haven't yet done that.